### PR TITLE
[6.x] Allow developer to preserve query parameters on paginated Api resources

### PR DESCRIPTION
**Pull request by deleugpn**
Reattempt at #30729 without static configuration.

Instead of statically configuring it once, the developer would have to call `->preserveQueryParameters()` after instantiating the Resource Collection, but I still like the DX with this case.

```
class MyHandler
{
    public function __invoke(MyRepository $repository)
    {
        return MyResourceCollection::make($repository->paginate())->preserveQueryParameters();
    }
}
```

### CODE REVIEW

File: src/Illuminate/Http/Resources/Json/ResourceCollection.php [hunk 1]
```diff
@@ -67,8 +86,25 @@ public function toArray($request)
...
+        return parent::toResponse($request);
+    }
+
+    /**
+     * Create an paginate-aware HTTP response.
+     *
+     * @param \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\JsonResponse
+     */
+    private function preparePaginatedResponse($request)
```
Comment (tobiasthaden) [inline src/Illuminate/Http/Resources/Json/ResourceCollection.php]: It's okay to have this as method as `protected`. I think.

Comment (deleugpn) [reply src/Illuminate/Http/Resources/Json/ResourceCollection.php]: Force of habit, I guess 😄

**Comment by taylorotwell** [comment]
Yeah this feels a lot better. Thanks.

File: src/Illuminate/Http/Resources/Json/ResourceCollection.php [hunk 2]
```diff
@@ -25,6 +25,13 @@ class ResourceCollection extends JsonResource implements Countable, IteratorAggr
      */
     public $collection;
 
+    /**
+     * Determines whether to preserve all query parameters when generating the navigation links.
+     *
+     * @var bool
+     */
+    private $queryParameters;
```
Comment (driesvints) [inline src/Illuminate/Http/Resources/Json/ResourceCollection.php]:
```suggestion
    protected $queryParameters;
```

File: src/Illuminate/Http/Resources/Json/ResourceCollection.php [hunk 3]
```diff
@@ -67,8 +86,25 @@ public function toArray($request)
...
+            return $this->preparePaginatedResponse($request);
+        }
+
+        return parent::toResponse($request);
+    }
+
+    /**
+     * Create an paginate-aware HTTP response.
+     *
+     * @param \Illuminate\Http\Request  $request
```
Comment (driesvints) [inline src/Illuminate/Http/Resources/Json/ResourceCollection.php]:
```suggestion
     * @param  \Illuminate\Http\Request  $request
```

File: src/Illuminate/Http/Resources/Json/ResourceCollection.php [hunk 4]
```diff
@@ -67,8 +86,25 @@ public function toArray($request)
...
-                    : parent::toResponse($request);
+        if ($this->resource instanceof AbstractPaginator) {
+            return $this->preparePaginatedResponse($request);
+        }
+
+        return parent::toResponse($request);
+    }
+
+    /**
+     * Create an paginate-aware HTTP response.
```
Comment (driesvints) [inline src/Illuminate/Http/Resources/Json/ResourceCollection.php]:
```suggestion
     * Create a paginate-aware HTTP response.
```

**Comment by driesvints** [comment]
Maybe rename the method to something less verbose? Thinking `withUrl` or something in the likes.

File: src/Illuminate/Http/Resources/Json/ResourceCollection.php [hunk 5]
```diff
@@ -25,6 +25,13 @@ class ResourceCollection extends JsonResource implements Countable, IteratorAggr
      */
     public $collection;
 
+    /**
+     * Determines whether to preserve all query parameters when generating the navigation links.
+     *
+     * @var bool
+     */
+    protected $queryParameters;
```
Comment (Alymosul) [inline src/Illuminate/Http/Resources/Json/ResourceCollection.php]: Maybe `shouldPreserveQueryParamters` would be better and expressive.

**Comment by taylorotwell** [comment]
I tweaked this into two methods and did some renaming:

image

**Comment by BrandonSurowiec** [comment]
Documentation will need to be added to laravel/docs for the new `preserveQuery()` and `withQuery()` methods.

**Pull request merged**